File: Kotlin/BlackJack/src/unitTests/PlayerTest.kt
package unitTests

import BlackJack.behavior.ClassicPlayer
import BlackJack.color.CardColor
import BlackJack.entity.Card
import BlackJack.entity.Player
import BlackJack.entity.TesterDealer
import BlackJack.input.ConsoleInputHandler
import BlackJack.type.CardType
import org.junit.Before
import org.junit.Test

import org.junit.Assert.*
import java.io.ByteArrayInputStream

class PlayerTest {

    private val player = Player(
            basicAmountOfMoney = 100.0,
            dealer = TesterDealer(1, TestDealerObserver()),
            behavior = ClassicPlayer(),
            observer = TestPlayerObserver(),
            inputHandler = ConsoleInputHandler()
    )

    @Before
    fun setUp() {

        val dealer = player.dealer as TesterDealer

        dealer.createNewDeck(
                arrayListOf(
                        Card(CardType.KING, CardColor.DIAMONDS),
                        Card(CardType.KING, CardColor.HEARTS),
                        Card(CardType.NINE, CardColor.CLUBS),
                        Card(CardType.FOUR, CardColor.CLUBS),
                        Card(CardType.QUEEN, CardColor.SPADES)

                )
        )

        player.changeDeck(0)

    }

    @Test
    fun doubleCheck() {

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        assertEquals(90.0, player.showMoney(), 0.01)

        player.double(player.showDeck())

        assertEquals(20.0, player.showDeck().money, 0.01)

        assertEquals(80.0, player.showMoney(), 0.01)

    }

    @Test
    fun multipleDoubleTest() {

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        assertEquals(90.0, player.showMoney(), 0.01)

        player.double(player.showDeck())
        player.double(player.showDeck())

        assertEquals(40.0, player.showDeck().money, 0.01)
        assertEquals(60.0, player.showMoney(), 0.01)

    }

    @Test
    fun checkMaximumDoubleTest() {

        System.setIn(ByteArrayInputStream("40".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        assertEquals(60.0, player.showMoney(), 0.01)

        player.double(player.showDeck())
        player.double(player.showDeck())

        assertEquals(100.0, player.showDeck().money, 0.01)
        assertEquals(0.0, player.showMoney(), 0.01)

    }

    @Test
    fun playAndQuitTest() {

        System.setIn(ByteArrayInputStream("40\ns\nd\nh\ne\nh\nq".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        player.play()

        assertNotEquals(100, player.showMoney())

    }

    @Test
    fun playAndNewGameTest() {

        System.setIn(ByteArrayInputStream("40\ns\nd\nh\ne\nh\nn\n500".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        player.play()

        assertNotEquals(500, player.showMoney())

    }

    @Test
    fun simpleHitTest() {

        val dealer = player.dealer as TesterDealer

        dealer.createNewDeck(
                arrayListOf(
                        Card(CardType.TEN, CardColor.DIAMONDS),
                        Card(CardType.THREE, CardColor.HEARTS),
                        Card(CardType.FOUR, CardColor.SPADES),
                        Card(CardType.ACE, CardColor.CLUBS),
                        Card(CardType.JACK, CardColor.DIAMONDS),
                        Card(CardType.ACE, CardColor.DIAMONDS)

                )
        )

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        player.hit(player.showDeck())

        assertEquals(3, player.showDeck().cards.count())

    }

    @Test
    fun multipleHitTest() {

        val dealer = player.dealer as TesterDealer

        dealer.createNewDeck(
                arrayListOf(
                        Card(CardType.TEN, CardColor.DIAMONDS),
                        Card(CardType.THREE, CardColor.HEARTS),
                        Card(CardType.FOUR, CardColor.SPADES),
                        Card(CardType.ACE, CardColor.CLUBS),
                        Card(CardType.JACK, CardColor.DIAMONDS),
                        Card(CardType.ACE, CardColor.DIAMONDS)

                )
        )

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        player.hit(player.showDeck())
        player.hit(player.showDeck())
        player.hit(player.showDeck())

        assertEquals(5, player.showDeck().cards.count())

    }

    @Test
    fun checkMaximumHitTest() {

        val dealer = player.dealer as TesterDealer

        dealer.createNewDeck(
                arrayListOf(
                        Card(CardType.TEN, CardColor.DIAMONDS),
                        Card(CardType.THREE, CardColor.HEARTS),
                        Card(CardType.FOUR, CardColor.SPADES),
                        Card(CardType.ACE, CardColor.CLUBS),
                        Card(CardType.JACK, CardColor.DIAMONDS),
                        Card(CardType.ACE, CardColor.DIAMONDS)

                )
        )

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        player.hit(player.showDeck())
        player.hit(player.showDeck())
        player.hit(player.showDeck())

        player.hit(player.showDeck())

        assertEquals(5, player.showDeck().cards.count())

    }

    @Test
    fun splitTwoCardsAndWithTheSameValueTest() {

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        player.split(player.showDeck())

        assertEquals(1, player.showDeck().cards.count())
        assertEquals(5.0, player.showDeck().money, 0.01)

        player.changeDeck(player.numberOfUsedDecks - 1)


        assertEquals(1, player.showDeck().cards.count())
        assertEquals(5.0, player.showDeck().money, 0.01)

        assertEquals(Card(CardType.KING, CardColor.DIAMONDS), player.showDeck().cards[0])
    }

    @Test
    fun splitWithTwoCardsButNotTheSameValueTest() {

        //A két ugyan olyan kártya kivétele (az egységes kezelés és a boiler text csökkentése miatt)
        player.dealer.giveCard()
        player.dealer.giveCard()

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        val currentUsedDeckNumber = player.numberOfUsedDecks

        player.split(player.showDeck())

        assertEquals(2, player.showDeck().cards.count())
        assertEquals(10.0, player.showDeck().money, 0.01)

        player.changeDeck(currentUsedDeckNumber + 1)


        assertEquals(0, player.showDeck().cards.count())
        assertEquals(0.0, player.showDeck().money, 0.01)

    }

    @Test
    fun splitButNotWithTwoNumberTest() {

        //A két ugyan olyan kártya kivétele (az egységes kezelés és a boiler text csökkentése miatt)
        player.dealer.giveCard()
        player.dealer.giveCard()

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        val currentUsedDeckNumber = player.numberOfUsedDecks

        player.hit(player.showDeck())

        player.split(player.showDeck())

        assertEquals(3, player.showDeck().cards.count())
        assertEquals(10.0, player.showDeck().money, 0.01)

        player.changeDeck(currentUsedDeckNumber + 1)


        assertEquals(0, player.showDeck().cards.count())
        assertEquals(0.0, player.showDeck().money, 0.01)

    }

    @Test
    fun multipleSplitAndPreparationTest() {

        val dealer = player.dealer as TesterDealer

        dealer.createNewDeck(
                arrayListOf(
                        Card(CardType.JACK, CardColor.DIAMONDS),
                        Card(CardType.JACK, CardColor.HEARTS),
                        Card(CardType.JACK, CardColor.SPADES),
                        Card(CardType.JACK, CardColor.CLUBS),
                        Card(CardType.JACK, CardColor.DIAMONDS)

                )
        )

        System.setIn(ByteArrayInputStream("10".toByteArray()))

        player.raise()
        player.receiveFirstCards()

        player.split(player.showDeck())

        assertEquals(1, player.showDeck().cards.count())
        assertEquals(5.0, player.showDeck().money, 0.01)


        player.hit(player.showDeck())

        player.split(player.showDeck())

        assertEquals(1, player.showDeck().cards.count())
        assertEquals(2.5, player.showDeck().money, 0.01)

        player.hit(player.showDeck())

        player.split(player.showDeck())

        assertEquals(2, player.showDeck().cards.count())
        assertEquals(2.5, player.showDeck().money, 0.01)

        player.changeDeck(player.numberOfUsedDecks - 1)

        assertEquals(1, player.showDeck().cards.count())
        assertEquals(2.5, player.showDeck().money, 0.01)

        // ---------------------------------------------------------
        //Itt talán a legjobb tesztelni a preparation függvényt
        // ---------------------------------------------------------

        (0 until player.numberOfUsedDecks).forEach {
            player.changeDeck(it)

            if(it == 0)
                assertEquals(2, player.showDeck().cards.count())
            else
                assertEquals(1, player.showDeck().cards.count())
        }

        assertEquals(3, player.numberOfUsedDecks)

        player.preparation()

        assertEquals(1, player.numberOfUsedDecks)

        (0 until player.behavior.getMaximumNumberOfDecks()).forEach {
            player.changeDeck(it)
            assertEquals(0, player.showDeck().cards.count())
            assertEquals(0.0, player.showDeck().money, 0.01)
        }


    }

    @Test
    fun raiseTest() {

        System.setIn(ByteArrayInputStream("1a\n100.1\n100".toByteArray()))

        player.raise()

        assertEquals(0.0, player.showMoney(), 0.01)
        assertEquals(100.0, player.showDeck().money, 0.01)

        (1 until player.behavior.getMaximumNumberOfDecks()).forEach {

            player.changeDeck(it)

            assertEquals(0.0, player.showDeck().money, 0.01)
        }

    }

    @Test
    fun receiveFirstCardsTest() {

        player.receiveFirstCards()

        assertEquals(Card(CardType.KING, CardColor.DIAMONDS), player.showDeck().cards[0])
        assertEquals(Card(CardType.KING, CardColor.HEARTS), player.showDeck().cards[1])


    }

}

File: Lua/BlackJack/blackjack/test/playerTest.lua

h = require "test.testHelper"
io.read = h.read

local Player = require "entity.Player"
local Dealer = require "entity.Dealer"
local ActionType = require "type.ActionType"
local GameState = require "type.GameState"
local ResultType = require "type.ResultType"
local ClassicTableView = require "gfx.ClassicTableView"
local ClassicPlayer = require "behavior.ClassicPlayer"
local ClassicBank = require "behavior.ClassicBank"
local Calculator = require("entity.Calculator"):getInstance()
local ConsoleInputHandler = require("input.ConsoleInputHandler"):getInstance()
local testPlayerObserver = require "test.testPlayerObserver"
local Card = require "entity.Card"

local testDealer = {
    deckPool = {}
}
function testDealer:giveCard()
    return table.remove(self.deckPool, 1)
end

function setUp(player) 

    player.dealer.deckPool = {
        Card("KING", "DIAMONDS", 11),
        Card("KING", "HEARTS", 11),
        Card("NINE", "CLUBS", 9),
        Card("FOUR", "CLUBS", 4),
        Card("QUEEN", "SPADES", 11)
    }

    player:changeDeck(0)

end


function testDoubleCheck()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"10"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()
    
    lu.assertEquals(player:showMoney(), 90.0)
    
    player:double(player:showDeck())
    
    lu.assertEquals(player:showDeck().money, 20.0)
    
    lu.assertEquals(player:showMoney(), 80.0)

end


function testMultipleDouble()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"10"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()
    
    lu.assertEquals(player:showMoney(), 90.0)

    player:double(player:showDeck())
    player:double(player:showDeck())
    
    lu.assertEquals(player:showDeck().money, 40.0)
    lu.assertEquals(player:showMoney(), 60.0)

end


function tetsCheckMaximumDouble()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"40"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()

    lu.assertEquals(player:showMoney(), 40.0)

    player:double(player:showDeck())
    player:double(player:showDeck())

    lu.assertEquals(player:showDeck().money, 100.0)
    lu.assertEquals(player:showMoney(), 0.0)

end


function testPlayAndQuit()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"40", "s", "d", "h", "e", "h", "q" }
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()

    player:play()

    lu.assertNotEquals(player:showMoney(), 100.0)

end


function testPlayAndNewGame()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"40", "s", "d", "h", "e", "h", "n", "500"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()

    player:play()

    lu.assertNotEquals(player:showMoney(), 500)

end

function setUp2(player)

    player.dealer.deckPool = {
        Card("TEN", "DIAMONDS", 10),
        Card("THREE", "HEARTS", 3),
        Card("FOUR", "SPADES", 4),
        Card("ACE", "CLUBS", 11),
        Card("JACK", "DIAMONDS", 11),
        Card("ACE", "DIAMONDS", 11)
    }

    player:changeDeck(0)

end


function testSimpleHit()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp2(player)
    h.input = {"10"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()

    player:hit(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 3)

end


function testMultipleHit()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp2(player)
    h.input = {"10"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()

    player:hit(player:showDeck())
    player:hit(player:showDeck())
    player:hit(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 5)

end


function testCheckMaximumHit()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp2(player)
    h.input = {"10"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()

    player:hit(player:showDeck())
    player:hit(player:showDeck())
    player:hit(player:showDeck())

    player:hit(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 5)

end


function testSplitTwoCardsAndWithTheSameValue()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"10"}
    h.cnt = 0

    player:raise()
    player:receiveFirstCards()

    player:split(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 1)
    lu.assertEquals(player:showDeck().money, 5.0)

    player:changeDeck(player.numberOfUsedDecks)


    lu.assertEquals(#player:showDeck().cards, 1)
    lu.assertEquals(player:showDeck().money, 5.0)

    lu.assertEquals(player:showDeck().cards[1].value, "KING")
    lu.assertEquals(player:showDeck().cards[1].color, "DIAMONDS")
end


function tetsSplitWithTwoCardsButNotTheSameValue()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"10"}
    h.cnt = 0

    --A két ugyan olyan kártya kivétele (az egységes kezelés és a boiler text csökkentése miatt)
    player.dealer:giveCard()
    player.dealer:giveCard()


    player:raise()
    player:receiveFirstCards()

    local currentUsedDeckNumber = player.numberOfUsedDecks

    player:split(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 2)
    lu.assertEquals(player:showDeck().money, 10.0)

    player:changeDeck(currentUsedDeckNumber + 1)


    lu.assertEquals(#player:showDeck().cards, 0)
    lu.assertEquals(player:showDeck().money, 0.0)

end


function testSplitButNotWithTwoNumber()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"10"}
    h.cnt = 0

    --A két ugyan olyan kártya kivétele (az egységes kezelés és a boiler text csökkentése miatt)
    player.dealer:giveCard()
    player.dealer:giveCard()

    player:raise()
    player:receiveFirstCards()

    local currentUsedDeckNumber = player.numberOfUsedDecks

    player:hit(player:showDeck())

    player:split(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 3)
    lu.assertEquals(player:showDeck().money, 10.0)

    player:changeDeck(currentUsedDeckNumber + 1)


    lu.assertEquals(#player:showDeck().cards, 0)
    lu.assertEquals(player:showDeck().money, 0.0)

end


function testMultipleSplitAndPreparation()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"10"}
    h.cnt = 0

    player.dealer.deckPool = {
        Card("JACK", "DIAMONDS", 11),
        Card("JACK", "HEARTS", 11),
        Card("JACK", "SPADES", 11),
        Card("JACK", "CLUBS", 11),
        Card("JACK", "DIAMONDS", 11)
    }

    player:raise()
    player:receiveFirstCards()

    player:split(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 1)
    lu.assertEquals(player:showDeck().money, 5.0)


    player:hit(player:showDeck())

    player:split(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 1)
    lu.assertEquals(player:showDeck().money, 2.5)

    player:hit(player:showDeck())

    player:split(player:showDeck())

    lu.assertEquals(#player:showDeck().cards, 2)
    lu.assertEquals(player:showDeck().money, 2.5)

    player:changeDeck(player.numberOfUsedDecks)

    lu.assertEquals(#player:showDeck().cards, 1)
    lu.assertEquals(player:showDeck().money, 2.5)

    -- ---------------------------------------------------------
    -- Itt talán a legjobb tesztelni a preparation függvényt
    -- ---------------------------------------------------------

    for i = 1, player.numberOfUsedDecks, 1 do

        player:changeDeck(i)
        if(i == 1) then lu.assertEquals(#player:showDeck().cards, 2)
        else
            lu.assertEquals(#player:showDeck().cards, 1)
        end

    end

    lu.assertEquals(player.numberOfUsedDecks, 3)

    player:preparation()

    lu.assertEquals(player.numberOfUsedDecks, 1)

    for i = 1, player.behavior:getMaximumNumberOfDecks(), 1 do

        player:changeDeck(i)
        lu.assertEquals(#player:showDeck().cards, 0)
        lu.assertEquals(player:showDeck().money, 0.0)

    end

end


function testRaise()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)
    h.input = {"1a", "100.1", "100"}
    h.cnt = 0

    player:raise()

    lu.assertEquals(player:showMoney(), 0.0)
    lu.assertEquals(player:showDeck().money, 100.0)

    for i = 2, player.behavior:getMaximumNumberOfDecks(), 1 do

        player:changeDeck(i)
        lu.assertEquals(player:showDeck().money, 0.0)

    end

end


function testReceiveFirstCards()

    local player = Player(100.0, testDealer, ClassicPlayer(), ConsoleInputHandler, testPlayerObserver)
    setUp(player)

    player:receiveFirstCards()

    lu.assertEquals("KING", player:showDeck().cards[1].value)
    lu.assertEquals("DIAMONDS", player:showDeck().cards[1].color)
    lu.assertEquals("KING", player:showDeck().cards[2].value)
    lu.assertEquals("HEARTS", player:showDeck().cards[2].color)


end
